lib/internal/backend/repository/repository2.go

```diff
@@ -294,49 +294,17 @@ func getLayerIndex(layerID string, manifest v2Manifest) (int, error) {
 }
 
 func (rb *RepositoryBackend) getLayerV2(layerID string, dockerURL *types.ParsedDockerURL, tmpDir string, copier *progressutil.CopyProgressPrinter) (*os.File, io.ReadCloser, error) {
-	url := rb.schema + path.Join(dockerURL.IndexURL, "v2", dockerURL.ImageName, "blobs", layerID)
-	req, err := http.NewRequest("GET", url, nil)
+	dockerImgSource, err := docker.NewDockerImageSource(dockerURL.IndexURL+"/"+dockerURL.ImageName+":"+dockerURL.Tag, "", false)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	rb.setBasicAuth(req)
-
-	res, err := rb.makeRequest(req, dockerURL.ImageName)
+	in, size, err := dockerImgSource.GetBlob(layerID)
 	if err != nil {
 		return nil, nil, err
 	}
-
-	if res.StatusCode == http.StatusTemporaryRedirect || res.StatusCode == http.StatusFound {
-		location := res.Header.Get("Location")
-		if location != "" {
-			req, err = http.NewRequest("GET", location, nil)
-			if err != nil {
-				return nil, nil, err
-			}
-			res, err = rb.makeRequest(req, dockerURL.ImageName)
-			if err != nil {
-				return nil, nil, err
-			}
-			defer res.Body.Close()
```

Comment:
this is probably a bug in `master` as I don't see how can you return a closed response body at the end :confused: